sql-plugin/src/main/spark340/scala/com/nvidia/spark/rapids/shims/LimitShims.scala

```diff
@@ -0,0 +1,100 @@
+      // offset is introduce in spark-3.4.0, and it ensures that offset >= 0. And limit can be -1,
+      // such case happens only when we execute sql like 'select * from table offset 10'
+      val offset = plan.wrapped.offset
+      val limit = plan.wrapped.limit
+      val sliced = if (limit >= 0) {
+        Some(BigInt(limit - offset).max(0))
+      } else {
+        // limit can be -1, meaning no limit
+        None
+      }
+      RowCountPlanVisitor.visit(plan.childPlans.head)
+          .map { rowNum =>
+            val remaining = (rowNum - offset).max(0)
+            sliced.map(_.min(remaining)).getOrElse(remaining)
+          }
+          .orElse(sliced)
+    }
+  }
+
+  def execs: Map[Class[_ <: SparkPlan], ExecRule[_ <: SparkPlan]] =
+    Seq(
+      exec[TakeOrderedAndProjectExec](
+        "Take the first limit elements as defined by the sortOrder, and do projection if needed",
+        // The SortOrder TypeSig will govern what types can actually be used as sorting key data
+        // type. The types below are allowed as inputs and outputs.
+        ExecChecks((pluginSupportedOrderableSig + TypeSig.DECIMAL_128 +
+            TypeSig.ARRAY + TypeSig.STRUCT + TypeSig.MAP).nested(), TypeSig.all),
+        (takeExec, conf, p, r) =>
+          new SparkPlanMeta[TakeOrderedAndProjectExec](takeExec, conf, p, r) {
+            val sortOrder: Seq[BaseExprMeta[SortOrder]] =
+              takeExec.sortOrder.map(GpuOverrides.wrapExpr(_, conf, Some(this)))
+            val projectList: Seq[BaseExprMeta[NamedExpression]] =
+              takeExec.projectList.map(GpuOverrides.wrapExpr(_, conf, Some(this)))
+            override val childExprs: Seq[BaseExprMeta[_]] = sortOrder ++ projectList
+
+            override def convertToGpu(): GpuExec = {
+              // To avoid metrics confusion we split a single stage up into multiple parts but only
+              // if there are multiple partitions to make it worth doing.
+              val so = sortOrder.map(_.convertToGpu().asInstanceOf[SortOrder])
+              if (takeExec.child.outputPartitioning.numPartitions == 1) {
+                GpuTopN(takeExec.limit, so,
+                  projectList.map(_.convertToGpu().asInstanceOf[NamedExpression]),
+                  childPlans.head.convertIfNeeded(), takeExec.offset)(takeExec.sortOrder)
+              } else {
+                GpuTopN(
+                  takeExec.limit,
+                  so,
+                  projectList.map(_.convertToGpu().asInstanceOf[NamedExpression]),
+                  GpuShuffleExchangeExec(
+                    GpuSinglePartitioning,
+                    GpuTopN(
+                      takeExec.limit,
+                      so,
+                      takeExec.child.output,
+                      childPlans.head.convertIfNeeded())(takeExec.sortOrder),
+                    ENSURE_REQUIREMENTS
+                  )(SinglePartition),
+                  takeExec.offset)(takeExec.sortOrder)
```

Comment:
nit: can we add a note to indicate that we are only doing the offset in GpuTopN on the last task.

Reply:
What do you mean by everything being in a single task? The If on line 77 checks to see if all of the data is already in a single task, and then does not need to insert in a shuffle.  The else, which is where this code is, will insert a shuffle after the first GpuTopN that goes to a `SinglePartition` that forces it to a single task so we can get the limit to work properly.

Reply:
We talked about this offline and resolved the confusion. Just to close the loop here, I was referring to the second branch (else case). We plan to address this better in a follow-on issue.